### ISSUE TITLE
local data not found when adding a torrent to a non-default folder

### ISSUE THREAD

**Issue by ilikepeaches** [transmission/transmission#4987]
### What is the issue?

If you tell Transmission to download a torrent's data to a _custom location_ (not the "Default location" specified in Settings), Transmission should check that location (and nowhere else) to see whether the data already exists.  Based on what it finds, it should then seed the data that's already there and download any data that isn't.

I don't think that I just wrote a feature request.  I think I just summarized expected behavior!  But that's not what Transmission actually does.

Instead, Transmission consistently & reproducibly gets this task wrong: reporting 0% data when it should see 100%; re-downloading entire torrents unnecessarily; overwriting or corrupting perfectly good data; and — here is a likely cause of the issue — seeming always to look in the "Default location," even when a custom one is specified.

---

### Data in default location + download to default location
**Expected: torrent loads green at 100%
Actual: torrent loads green at 100%**
1. You already have a torrent's data.
2. The "Default location" is specified in Settings: ~/Downloads/default-location/

![00  default location config](https://user-images.githubusercontent.com/44431584/220639366-3b2b2aa9-34dc-431d-b225-a13469355ceb.png)

3. The data is in the "Default location" directory: ~/Downloads/default-location/

![00a  data in default location](https://user-images.githubusercontent.com/44431584/220639404-c6746a2c-7684-40b6-b52b-058b69e44ec3.png)

4. You add the torrent to Transmission, either specifying the "Default location" or allowing it to be chosen automatically:  ~/Downloads/default-location/

![01  add torrent to default location](https://user-images.githubusercontent.com/44431584/220639437-ad40df2b-d9b7-4b9f-a96f-2933d7a0b1c4.png)

Notice the "100% of 403.7 MB" shown here.

5. Transmission recognizes that you already have the data, fast verifies, and loads the torrent at 100% & green.

![02  torrent adds green yay](https://user-images.githubusercontent.com/44431584/220639479-74c663f7-ac7a-4baa-bc76-15560eeda90a.png)

So far, so good!

---

### Data in custom location + download to custom location
**Expected: torrent loads green at 100%
Actual: torrent loads red at 0%**
1. You already have a torrent's data.
2. The "Default location" is specified in Settings: ~/Downloads/default-location/
3. The data is in a **custom location** directory: ~/Downloads/custom-location/
 
![02z](https://user-images.githubusercontent.com/44431584/220639522-93c134e1-64cc-44f2-b33c-109691ab8b4a.png)

4. You add the torrent to Transmission, specifying the custom location or allowing it to be chosen automatically e.g. by group rules:  ~/Downloads/custom-location/

![03  custom group location config](https://user-images.githubusercontent.com/44431584/220639551-68eb9a71-d605-4bad-b21d-3f739c408261.png)
![04  add torrent to custom location](https://user-images.githubusercontent.com/44431584/220639582-b796384b-dcd3-462b-8d12-ebec70af0a56.png)

Notice the "0% of 403.7 MB" shown here.

5. Transmission does NOT recognize that you already have the data, adding the torrent at 0% and red.

![05  torrent adds 0% boo](https://user-images.githubusercontent.com/44431584/220639606-cedb8cad-f0e3-4a84-a795-834b40fbf8c7.png)

7. If you click the magnifying glass to show the red torrent's data, Transmission will open ~/Downloads/custom-location/ in the Finder, showing you the 100% data you already have but that it has failed to recognize.
8. If you don't intercede, Transmission will begin re-downloading the torrent, overwriting (this is bad!) the 100% data in ~/Downloads/custom-location/ with partial data until the torrent completes.
9. If you manage to pause the torrent before re-downloading begins, you can "Verify local data," and Transmission will grind your hard drive for a period of time before confirming that actually, oops, you do indeed already have 100% of the data, right where you said it would be...

---

### Data in default location + download to custom location
**Expected: torrent loads at 0%
Actual: torrent loads at 100%, but gray & errored**
1. You already have a torrent's data.
2. The "Default location" is specified in Settings: ~/Downloads/default-location/
3. The data is in the "Default location" directory: ~/Downloads/default-location/

![07  data in default location](https://user-images.githubusercontent.com/44431584/220639657-419eace1-080a-47f9-88b9-85d9993dccb5.png)

4. You add the torrent to Transmission, specifying the **custom location** or allowing it to be chosen automatically e.g. by group rules:  ~/Downloads/custom-location/

![08  add torrent to custom location](https://user-images.githubusercontent.com/44431584/220639686-5b142b4c-4965-4a6e-8140-edc447b56410.png)

Notice the "100% of 403.7 MB" shown here.

5. Transmission adds the torrent at 100% and gray.

![09  torrent adds 100% gray boo](https://user-images.githubusercontent.com/44431584/220639719-5a613ddd-a7c0-4bd1-acf6-d7eba4683146.png)

The Activity pane of the Torrent Inspector shows "Error: No data found! Ensure your drives are connected or use 'Move Data File To…'. To re-download, remove the torrent and re-add it."

![10  error in inspector](https://user-images.githubusercontent.com/44431584/220639747-534a33c7-1122-46ae-9c45-ef439464aab6.png)

If you try to start the torrent, it will not start.  Transmission seems to recognize that you do have the data, which it shouldn't (because the data isn't in the location you specified) and also at the same time thinks you don't have the data, which it's mostly right about (because the data isn't in the location you specified).

---

Transmission 4's excellent new fast-verify feature (https://github.com/transmission/transmission/pull/2626) made this problem much easier to test.  In turn, this issue significantly undermines that feature.  I actually got here because I couldn't get 4.0.0's fast-verify behavior to work at all!  Eventually, I figured out the reason: all of my group data locations are custom, and fast-verify doesn't work if custom locations are involved.  It should, though.

I've done my testing & written this issue with the new version, ~~but this bug seems to predate Transmission 4.  For years, I've been adding torrents paused, seeing 0% even though I specified the correct 100% data location, then verifying that data up from 0 to 100% (R.I.P. HDDs), before resuming the torrent...~~

---

I can reproduce this issue 100% of the time on Mac Transmission 4.0.0-dev (9d1a141482), and another user was able to verify it on the Mac 4.0.0 stable release.  I don't know whether it affects more than the Mac GUI app but suspect that it does.  It's not just about custom locations _for groups_ (which are a Mac-only feature).  It happens any time you specify a custom location when adding a torrent.

So I hope that others can test it with the daemon and QT/GTK clients.

Thank you for reading :-)

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.0-dev (9d1a141)

**Comment by ckerr** [comment]
I wish all bug reports were this informative.

@nevack would you be interested in this one?

**Comment by ckerr** [comment]
@Coeur could you take a look at this one?

**Comment by Coeur** [comment]
I can confirm the 100%/0% disparity depending on whether it's a default folder or not. But that disparity was already happening in Transmission 3, so it's not completely a regression, even though we do want to address that issue.

**Comment by ilikepeaches** [comment]
https://github.com/transmission/transmission/pull/5060 forces a full verify every time.  We want fast-verify to work with existing data in custom locations.

**Comment by Coeur** [comment]
I can look at the "add torrent dialog".
But for the full verify / fast-verify, it does simply what you configured your client to do, so it's behaving as expected.

**Comment by Coeur** [comment]
I couldn't find your "add torrent" dialog.
All I found is:
- double-clic or drag&drop a torrent file (which is what my changes are covering explicitly)
- "Create Torrent file..."
- "Open Torrent file..."
- "Open Torrent address..."

**Comment by ilikepeaches** [comment]
Hmm.  My understanding was that fast-verify was on by default and was only configurable in the settings.json / org.m0k.transmission.plist.

Regardless, I have
```
	<key>torrent-added-verify-mode</key>
	<string>fast</string>
```
in the plist.

Further testing tonight with 4.0.1-dev (13ae244eed) (Mac/Intel):

- Fast-verify works as expected when the data location and default locations match.
- When the data is in a custom location, **and** Transmission sets the custom location via the Group preferences, then the data **full verifies** automatically from the add torrent window.
- When the data location is in a custom location, but there is no custom group location assigned, i.e. the default location is shown as the destination in the add torrent window, "Change..."-ing the location to the custom location should trigger a fast-verify, but it neither fast-verifies nor full verifies.

**Comment by ilikepeaches** [comment]
Oh shit it's not committed?  I swear it was ... I must've been looking in another branch?  I thought it'd been rolled into the latest nightlies.

That's my fault.  I will wait to test until it's actually testable!

in re: the add torrents window:

![Screenshot 2023-02-27 at 12 18 35 AM](https://user-images.githubusercontent.com/44431584/221480713-ca11239d-ac75-4035-9b66-f056d79ab26d.jpg)

gets you this on torrent add:

![Screenshot 2023-02-27 at 12 21 27 AM](https://user-images.githubusercontent.com/44431584/221480942-b2069395-2fbd-4424-882b-4dec514a0aa9.jpg)

**Comment by ilikepeaches** [comment]
Fixed in 4.0.1-dev (63653b4831) (Mac/Intel)!  Thank you!
